### PR TITLE
Add support for scanning Zstandard (zstd) compressed files

### PR DESCRIPTION
**Pull request by burakemir**
zstd is a fast compression library and associated format https://en.wikipedia.org/wiki/Zstd

ClamAV did not previously detect or decompress zstd-compressed files. This adds full support using the streaming ZSTD_decompressStream() API, following the same pattern as gzip/bzip2/xz decompression.

Changes:
- Magic byte detection (28 b5 2f fd) and CL_TYPE_ZSTD file type
- ARCH_CONF_ZSTD dconf flag for enable/disable control
- cli_scanzstd() streaming decompression scanner
- FindZSTD.cmake module and libzstd linkage
- clamscan test with zstd-compressed test fixture
- Document libzstd as a required dependency in INSTALL.md

This whole PR was done by Claude, I asked to add a test and ran it - take it or leave it!
I ran with MAINTAINER_MODE=ON to check that bindgen produces the same renumbering in "sys.rs".

### CODE REVIEW

**Comment by burakemir** [comment]
@codex review

File: libclamav/scanners.c [hunk 1]
```diff
@@ -1420,6 +1421,122 @@ static cl_error_t cli_scanxz(cli_ctx *ctx)
...
+                    cli_warnmsg("cli_scanzstd: decompress file size exceeds limits - "
+                                "only scanning %lu bytes\n",
+                                size);
+                    goto zstd_scan;
+                }
+            }
+
+            if (ores == 0) {
+                /* frame complete */
+                goto zstd_scan;
```
[the remark this answers is not in the document]
Comment (burakemir) [reply libclamav/scanners.c]: Done.

File: libclamav/scanners.c [hunk 2]
```diff
@@ -1420,6 +1421,122 @@ static cl_error_t cli_scanxz(cli_ctx *ctx)
...
+            ZSTD_outBuffer out;
+            out.dst  = obuf;
+            out.size = obuf_size;
+            out.pos  = 0;
+
+            ores = ZSTD_decompressStream(dstrm, &out, &in);
+            if (ZSTD_isError(ores)) {
+                cli_dbgmsg("cli_scanzstd: decompress error: %s\n", ZSTD_getErrorName(ores));
+                ret = CL_EFORMAT;
+                goto zstd_exit;
```
[the remark this answers is not in the document]
Comment (burakemir) [reply libclamav/scanners.c]: Done.

File: libclamav/scanners.c [hunk 3]
```diff
@@ -1420,6 +1421,122 @@ static cl_error_t cli_scanxz(cli_ctx *ctx)
...
+
+    do {
+        size_t avail;
+        ZSTD_inBuffer in;
+        const void *next_in = fmap_need_off_once_len(ctx->fmap, off, ibuf_size, &avail);
+
+        if (!next_in || avail == 0) {
+            if (size == 0) {
+                cli_dbgmsg("cli_scanzstd: premature end of compressed stream\n");
+                ret = CL_EFORMAT;
```
[the remark this answers is not in the document]
Comment (burakemir) [reply libclamav/scanners.c]: Done.

**Comment by burakemir** [comment]
Addressed the comments.

@codex review

File: libclamav/scanners.c [hunk 4]
```diff
@@ -1420,6 +1421,118 @@ static cl_error_t cli_scanxz(cli_ctx *ctx)
...
+        if (!next_in || avail == 0) {
+            break;
+        }
+        off += avail;
+
+        in.src  = next_in;
+        in.size = avail;
+        in.pos  = 0;
+
+        while (in.pos < in.size) {
```
[the remark this answers is not in the document]
Comment (burakemir) [reply libclamav/scanners.c]: Done.

**Comment by burakemir** [comment]
@codex review

File: cmake/FindZSTD.cmake [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+
+if(ZSTD_LIBRARY)
+  set(ZSTD_LIBRARIES "${ZSTD_LIBRARY}")
+endif()
+
+if(NOT ZSTD_LIBRARIES)
+  find_library(ZSTD_LIBRARY_RELEASE
+    NAMES zstd NAMES_PER_DIR ${PC_ZSTD_LIBRARY_DIRS} PATH_SUFFIXES lib)
+  find_library(ZSTD_LIBRARY_DEBUG
+    NAMES zstdd NAMES_PER_DIR ${PC_ZSTD_LIBRARY_DIRS} PATH_SUFFIXES lib)
```
[the remark this answers is not in the document]
Comment (burakemir) [reply cmake/FindZSTD.cmake]: Done.

**Comment by opoplawski** [comment]
conda packages are zip archives containing zstandard compressed tar files.  It would be great if clamav could automatically scan the contents.

**Comment by val-ms** [comment]
@burakemir Another option is to pull in the https://crates.io/crates/zstd-pure-rs or https://crates.io/crates/ruzstd rust crate and do this from Rust. Implementing new features in Rust is the preferred direction for the project in part because it reduces headaches of bringing in C/C++ based library dependencies, which as you are no doubt aware is not a lot of fun.

**Comment by burakemir** [comment]
Thanks @val-ms — I missed this option, and I agree the Rust route is the better plan. It also fits the existing pattern, since libclamav already does bzip2 (`bzip2-rs`), inflate, and LHA (`delharc`) as pure-Rust decoders in `libclamav_rust`.

Plan to switch to a pure-Rust implementation:

- Add `ruzstd` to `libclamav_rust/Cargo.toml` (decode-only, which is all a scanner needs).
- Implement the streaming decompress + scan in `libclamav_rust/src/scanners.rs`, exposed via cbindgen and called from `scanners.c` for `CL_TYPE_ZSTD`, mirroring the existing bzip2 path.
- Drop `cmake/FindZSTD.cmake`, the `find_package(ZSTD)` wiring, the `ZSTD::zstd` linkage, and the INSTALL.md dependency note, so there's no new system dependency.
- Keep the magic-byte detection, `CL_TYPE_ZSTD`, `ARCH_CONF_ZSTD` dconf toggle, and the existing clamscan tests (single-frame + concatenated-frame fixtures) unchanged.

I'll verify `ruzstd` handles concatenated and skippable frames before pushing the reworked version.

**Comment by burakemir** [comment]
Reworked to use the pure-Rust `ruzstd` decoder instead of libzstd (pushed in 821530e53).

**What changed**
- New `cli_scanzstd` implemented in `libclamav_rust/src/scanners.rs` using `ruzstd` (decode-only), mirroring the existing `scan_lha_lzh` path: read from the fmap, stream-decompress each frame, scan via `magic_scan`. The C-side dispatch is unchanged and now resolves to the Rust function via cbindgen.
- Removed all libzstd plumbing: the C `cli_scanzstd()`, the `<zstd.h>` include, `find_package(ZSTD)`, the `ZSTD::zstd` linkage, `cmake/FindZSTD.cmake`, and the INSTALL.md dependency notes. No new system dependency. (This also resolves the earlier vcpkg/Windows `REQUIRED` concern.)
- Kept `CL_TYPE_ZSTD`, magic-byte detection, the `ARCH_CONF_ZSTD` dconf flag, and the existing tests.

**Corner cases carried over from the earlier C review**, all verified in the Rust version:
- Concatenated frames — decoder is recreated per frame until input is exhausted (`StreamingDecoder` decodes a single frame).
- Skippable frames — handled via the `SkipFrame` header error.
- Partial output is always scanned — on decode error, scan-limit hit, *or* a decoder panic (caught so it can't unwind across the FFI boundary), whatever was decompressed so far is still scanned, so appending junk after a payload isn't an evasion path.
- Empty frames decode cleanly (no false `CL_EFORMAT`); scan limits enforced between and within frames; skip-length and size counters can't overflow.

The `ZSTD_DCtx_reset()` ignored-return-value note no longer applies (a fresh decoder per frame; no reset call).

**Tests:** `clamscan/zstd_test.py` passes for both the single-frame and concatenated-frame fixtures, producing byte-identical decompressed output (matching hashes) to the previous implementation.

One note: ruzstd is decode-only, which is all a scanner needs. The Rust path buffers decompressed output in memory rather than a temp file, bounded by `check_scan_limits` between chunks — consistent with the existing LHA Rust scanner.

**Review by val-ms — CHANGES_REQUESTED** [review]
I found one correctness issue in the final Rust version.

Finding:
- `filetypes_int.h` only detects zstd streams that start with the regular frame magic `28 b5 2f fd`. A valid zstd stream may start with a skippable frame (`0x184D2A50` through `0x184D2A5F`, little-endian bytes `50..5f 2a 4d 18`). In that case ClamAV recognizes the file as binary data and never dispatches to `cli_scanzstd()`, even though the Rust scanner handles `SkipFrame` once called. I reproduced this by prefixing the existing zstd fixture with a skippable frame: `zstd -dc` outputs the test payload, but `clamscan -d unit_tests/input/other_sigs/zstd.hdb` returns `OK`.

Proposed file type rules:

```c
"0:0:28b52ffd:Zstandard compressed file:CL_TYPE_ANY:CL_TYPE_ZSTD:76",
"0:0:5?2a4d18:Zstandard skippable frame:CL_TYPE_ANY:CL_TYPE_ZSTD:76",
```

`5?2a4d18` covers the little-endian skippable-frame magic range. Please also add a test fixture where a skippable frame precedes a normal zstd frame.

The rest of the Rust scanner and wiring look good to me. Please squash the PR into a single signed commit when updating.

**Review by val-ms — CHANGES_REQUESTED** [review]
Thanks for the update. I apologize for the bad proposed rule in my previous review: the `0:0:5?2a4d18` form was my suggestion, and it was wrong.

Mode `0` filetype entries are decoded as raw hex with `CLI_MPOOL_HEX2STR`, so the `?` wildcard makes the built-in filetype database fail to load. Please change the skippable-frame rule in `libclamav/filetypes_int.h` to the wildcard-capable mode `1` form:

```c
"1:0:5?2a4d18:Zstandard skippable frame:CL_TYPE_ANY:CL_TYPE_ZSTD:76",
```

I verified this local change with `make install && ctest -V`; the full suite passed, including `clamscan/zstd_test.py::TC::test_zstd_skippable_prefix`. Everything else in the PR still looks good to me.

**Comment by burakemir** [comment]
Thanks for diagnosing the problem. I had seen the test failures, but I could not make sense of what was going on (and there was a passing libclamav_rust test). 

Now, with the line updated as requested, I still do get an error in Test 3 clamscan test_sigtool_02_rust_logs_messages_work ... I cannot quite make sense of this either.

**Comment by val-ms** [comment]
@burakemir Everything passed in my local test and the github actions test. I'll run it through jenkins now to see how that goes. If you're seeing a test failure on your side, perhaps a clean build would resolve it. I'm not familiar with that specific test failing anywhere else, though we do have valgrind test failures for freshclam and clamd on some platforms.

**Review by val-ms — APPROVED** [review]

**Comment by val-ms** [comment]
It looks great! Thank you @burakemir !
